inssqs/inssqs.go

```diff
@@ -0,0 +1,402 @@
+package inssqs
+
+import (
+	"context"
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/retry"
+	awsconfig "github.com/aws/aws-sdk-go-v2/config"
+	awssqs "github.com/aws/aws-sdk-go-v2/service/sqs"
+	"github.com/aws/aws-sdk-go-v2/service/sqs/types"
+	"github.com/aws/smithy-go/middleware"
+	"github.com/pkg/errors"
+	"github.com/useinsider/go-pkg/inslogger"
+	"github.com/useinsider/go-pkg/inssqs/sqs"
+	"sync"
+)
+
+type Interface interface {
+	SendMessageBatch(entries []SQSMessageEntry) (failed []SQSMessageEntry, err error)
+	DeleteMessageBatch(entries []SQSDeleteMessageEntry) (failed []SQSDeleteMessageEntry, err error)
+}
+
+type queue struct {
+	client   sqs.API
+	name     string
+	url      *string
+	endpoint string
+
+	retryCount int
+
+	maxBatchSize      int
+	maxBatchSizeBytes int
+	workers           int
+
+	logger *inslogger.AppLogger
+}
+
+// Config represents the configuration settings required for initializing an SQS queue.
+type Config struct {
+	Region            string // AWS region where the SQS queue resides.
+	QueueName         string // Name of the SQS queue.
+	RetryCount        int    // Number of retry attempts allowed for queue operations.
+	MaxBatchSize      int    // Maximum size of a message batch.
+	MaxBatchSizeBytes int    // Maximum size of a message batch in bytes.
+	MaxWorkers        int    // Maximum number of workers for concurrent operations.
+	LogLevel          string // Log level for SQS operations.
+
+	EndpointUrl string // Endpoint URL for AWS operations.
+}
+
+func NewSQS(config Config) Interface {
+	if config.Region == "" {
+		panic(ErrRegionNotSet)
+	}
+
+	if config.QueueName == "" {
+		panic(ErrQueueNameNotSet)
+	}
+
+	config.setDefaults()
+
+	cfg, err := awsconfig.LoadDefaultConfig(context.Background(),
+		awsconfig.WithRegion(config.Region),
+		awsconfig.WithRetryMaxAttempts(config.RetryCount),
+		awsconfig.WithRetryMode(aws.RetryModeAdaptive))
+	if err != nil {
+		panic(errors.Wrap(err, "error while loading aws sqs config"))
+	}
+
+	// set endpoint url if provided
+	if config.EndpointUrl != "" {
+		cfg.BaseEndpoint = aws.String(config.EndpointUrl)
+	}
+
+	var logger *inslogger.AppLogger
+	if config.LogLevel == "" {
+		logger = inslogger.NewNopLogger()
+	} else {
+		logger = inslogger.NewLogger(inslogger.LogLevel(config.LogLevel))
```

Comment:
I believe not depending on a specific logger implementation would be beneficial. Just printing out request ids and simpler unit testing comes to mind.

Reply:
ne oneriyosun burada tam olarak anlamadim. logger'i iptal edip std out yapmak mi

Reply:
Depending on interfaces rather than concrete implementations would be a good  idea.

Reply:
hmm makes sense

Reply:
done